core/trino-main/src/main/java/io/trino/operator/exchange/PartitioningExchanger.java

```diff
@@ -97,6 +86,21 @@ public synchronized void accept(Page page)
         }
     }
 
+    private Page extractPartitioningChannels(Page inputPage)
+    {
+        // hash value is pre-computed, only needs to extract that channel
+        if (hashChannel.isPresent()) {
+            return new Page(inputPage.getBlock(hashChannel.get()));
+        }
+
+        // extract partitioning channels
+        Block[] blocks = new Block[partitioningChannels.size()];
+        for (int i = 0; i < partitioningChannels.size(); i++) {
+            blocks[i] = inputPage.getBlock(partitioningChannels.get(i));
+        }
+        return new Page(inputPage.getPositionCount(), blocks);
```

Comment:
This will copy the blocks right ? We could avoid it using `Page#getColumns` (partitioningChannels could be int array)

Reply:
Just an array. Not blocks data. If done once per page the perfomance impact should be neglegible.